fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java

```diff
@@ -0,0 +1,420 @@
+     *
+     * @param groupKey group key
+     * @param timestamp timestamp
+     * @return partition metadata future
+     * @throws RecordCoreException if no partition is found
+     */
+    @Nonnull
+    private CompletableFuture<LucenePartitionInfoProto.LucenePartitionInfo> getPartitionInfoOrFail(@Nonnull Tuple groupKey, long timestamp) {
+        return assignPartitionInternal(groupKey, timestamp, false);
+    }
+
+    /**
+     * assign a partition for a document insert or delete.
+     *
+     * @param groupKey group key
+     * @param timestamp document timestamp
+     * @param createIfNotExists if no suitable partition is found for this timestamp,
+     *                          create when <code>true</code>. This parameter should be set to <code>true</code> when
+     *                          inserting a document, and <code>false</code> when deleting.
+     * @return partition metadata future
+     * @throws RecordCoreException if <code>createIfNotExists</code> is <code>false</code> and no suitable partition is found
+     */
+    @Nonnull
+    private CompletableFuture<LucenePartitionInfoProto.LucenePartitionInfo> assignPartitionInternal(@Nonnull Tuple groupKey, long timestamp, boolean createIfNotExists) {
+        Range range = new Range(state.indexSubspace.subspace(Tuple.from(groupKey, PARTITION_META_SUBSPACE)).pack(),
+                                state.indexSubspace.subspace(Tuple.from(groupKey, PARTITION_META_SUBSPACE, timestamp)).pack());
+
+        final AsyncIterable<KeyValue> rangeIterable = state.context.ensureActive().getRange(range, 1, true, StreamingMode.WANT_ALL);
+
+        return AsyncUtil.collect(rangeIterable).thenComposeAsync(targetPartition -> {
+            if (targetPartition.isEmpty()) {
+                return getOldestPartition(groupKey).thenApply(oldestPartition -> {
+                    if (oldestPartition == null) {
+                        if (!createIfNotExists) {
+                            throw new RecordCoreException("Partition metadata not found", LogMessageKeys.PARTITIONING_TIMESTAMP, timestamp);
+                        } else {
+                            return newPartitionMetadata(timestamp);
+                        }
+                    } else {
+                        return oldestPartition;
+                    }
+                });
+            } else {
+                return CompletableFuture.completedFuture(partitionInfoFromKV(targetPartition.get(0)));
+            }
+        });
+    }
+
+    /**
+     * get the <code>long</code> timestamp value from a {@link FDBIndexableRecord}, given a field name.
+     *
+     * @param fieldName field name
+     * @param rec record
+     * @return long if field is found
+     * @throws RecordCoreException if no field of type <code>long</code> with given name is found
+     */
+    @Nonnull
+    private <M extends Message> Long getPartitioningTimestampValue(@Nonnull String fieldName, @Nonnull FDBIndexableRecord<M> rec) {
```

Comment:
It feels like perhaps this should build (and cache) a `KeyExpression` to do this work, rather than calling `Message.getAllFields` all the time. Or at least do the `getDescriptorForType` and `findFieldByName` and `getField` itself.

I don't have a problem with this being in a later pull request, though.